library/alloc/src/collections/btree/map.rs

```diff
@@ -1914,9 +1928,15 @@ impl<K, V> FusedIterator for RangeMut<'_, K, V> {}
 #[stable(feature = "rust1", since = "1.0.0")]
 impl<K: Ord, V> FromIterator<(K, V)> for BTreeMap<K, V> {
     fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> BTreeMap<K, V> {
-        let mut map = BTreeMap::new();
-        map.extend(iter);
-        map
+        let mut inputs: Vec<_> = iter.into_iter().collect();
```

Comment:
could this allocation be avoided by using a fixed-sized buffer on the stack instead and deduping/inserting in batches?

Reply:
Otherwise this allocation might trigger an OOM when streaming more duplicate tuples from disk than would fit into memory.

Reply:
This would a rare but hard to solve edge case. I can think two possible solutions:

* Fallback to use `extend`.
```rust
let iter = iter.into_iter();
let mut inputs: Vec<_> = (&mut iter).take(MAX_BUF_SIZE).collect();
inputs.sort_by(|a, b| a.0.cmp(&b.0));
let mut map = BTreeMap::bulk_build_from_sorted_iter(inputs.into_iter());
map.extend(iter);
map
```
This would lead to suboptimal performance. Another issue is what `MAX_BUF_SIZE` should be.

* Use Merge sort like algorithm.
```rust
let iter = iter.into_iter();
let mut buf_len: usize = INIT_BUF_SIZE;
let mut left_buf: Vec<_> = (&mut iter).take(buf_len).collect();
left_buf.sort_by(|a, b| a.0.cmp(&b.0));
let mut right_buf: Vec<_> = (&mut iter).take(buf_len).collect();
right_buf.sort_by(|a, b| a.0.cmp(&b.0));

while !right_buf.is_empty() {
    left_buf = MergeIter(MergeIterInner::new(
        DedupSortedIter::new(left_buf.into_iter()),
        DedupSortedIter::new(right_buf.into_iter()),
    ))
    .collect();

    buf_len = left_buf.len(); // in normal case this would increase exponentially
    right_buf = (&mut iter).take(buf_len).collect();
    right_buf.sort_by(|a, b| a.0.cmp(&b.0));
}

// in the end, the left_buf should have the same size as all non-dup keys in iter.
BTreeMap::bulk_build_from_sorted_iter(left_buf.into_iter())
```
This solution is more complex, but should achieve better performance in normal case. In cases there are lots of duplicates, the buffer size would not increase by much. There is a minor issue that how should we choose  `INIT_BUF_SIZE`.

Comments?

Reply:
I just tested my second solution. The performance is not that good. The `INIT_BUF_SIZE` is set to be 64.
```
test btree::map::from_iter_rand_100                      ... bench:       5,457 ns/iter (+/- 819)
test btree::map::from_iter_rand_10_000                   ... bench:   1,057,839 ns/iter (+/- 138,483)
test btree::map::from_iter_seq_100                       ... bench:       5,660 ns/iter (+/- 739)
test btree::map::from_iter_seq_10_000                    ... bench:     672,612 ns/iter (+/- 100,751)
```

Reply:
I had the following approach in mind

1. fill vec up to `MAX_BUF_SIZE`
2. sort
3. `bulk_build_from_sorted_iter`
4. if iterator is not empty goto 1

Reply:
Oh wait, we don't have an extend method that is optimized for receiving a sorted batch.

Then it might be possible to improve your second approach by reusing allocations. There are too many `collect` calls.
And `INIT_BUF_SIZE` could be chosen based on the iterator's size hint.

Reply:
> Oh wait, we don't have an extend method that is optimized for receiving a sorted batch.

We actually have one, i.e., `BTreeMap::append`. However, the performance would be even worse as it works by rebuilding the btree from the sorted iter.

> Then it might be possible to improve your second approach by reusing allocations

The issue is that each `collect` calls have different buf length, which increases exponentially. I am not sure how you can reuse the allocations here.

Reply:
> I am not sure how you can reuse the allocations here.

If I understood the approach correctly then it's not necessary for `right_buf` to grow much when there are many duplicates.

> We actually have one, i.e., BTreeMap::append. However, the performance would be even worse as it works by rebuilding the btree from the sorted iter.

Yeah, we would need something more efficient that only allocates the extra nodes needed and then redistribute entries from the map and the iterator.

----

Anyway, looking at the benchmarks it looks like you're only covering cases with few duplicates. Measuring an iterator with many duplicates might show different performance characteristics since you may end up allocating a large vec where before it would only need to insert a few elements into the map.

Reply:
I don't think we necessarily need to bother with the OOM condition here, we can potentially do that in a follow up PR.

Reply:
> Anyway, looking at the benchmarks it looks like you're only covering cases with few duplicates. Measuring an iterator with many duplicates might show different performance characteristics since you may end up allocating a large vec where before it would only need to insert a few elements into the map.

My sense of typical allocation performance (though not really backed up by anything) is that generally the size doesn't matter *that* much -- you're either pulling from a pool for tiny sizes or likely falling back to mmap or similar for larger allocations. So I wouldn't expect much difference here.

I'd guess that the bulk loading is going to be pretty efficient and in most cases you're not building BTreeMap and similar from iterators with lots of duplicates, I suspect. If we optimize for that case we'll probably hurt the performance of the low-duplicates/no-duplicates case. In practice, we're going to allocate `length / 2` buffer for the merge sort *anyway* and 2x difference in allocation size probably won't matter that much.